be/src/olap/rowset/segment_v2/variant/variant_column_writer_impl.cpp

```diff
@@ -366,6 +388,114 @@ Status UnifiedSparseColumnWriter::append_from_variant(const vectorized::ColumnVa
     return Status::OK();
 }
 
+Status VariantDocSnapShotWriter::init(const TabletColumn* parent_column, int bucket_num,
+                                      int& column_id, const ColumnWriterOptions& opts,
+                                      SegmentFooterPB* footer) {
+    _bucket_num = bucket_num;
+    _first_column_id = column_id;
+    _doc_snapshot_column_writers.resize(_bucket_num);
+    _doc_snapshot_column_opts.resize(_bucket_num);
+    for (int b = 0; b < _bucket_num; ++b) {
+        const TabletColumn& bucket_column =
+                vectorized::schema_util::create_doc_snapshot_column(*parent_column, b);
+        _doc_snapshot_column_opts[b] = opts;
+        _doc_snapshot_column_opts[b].meta = footer->add_columns();
+        _init_column_meta(_doc_snapshot_column_opts[b].meta, column_id, bucket_column,
+                          opts.compression_type);
+        RETURN_IF_ERROR(ColumnWriter::create_map_writer(_doc_snapshot_column_opts[b],
+                                                        &bucket_column, opts.file_writer,
+                                                        &_doc_snapshot_column_writers[b]));
+        RETURN_IF_ERROR(_doc_snapshot_column_writers[b]->init());
+        ++column_id;
+    }
+    return Status::OK();
+}
+
+Status VariantDocSnapShotWriter::append_data(const TabletColumn* parent_column,
```

Comment:
need more comment， maybe some logic could be reused with UnifiedSparseColumnWriter